ember-resources/src/util/function.ts

```diff
@@ -138,36 +139,76 @@ export function trackedFunction<Return>(...passedArgs: UseFunctionArgs<Return>)
  * State container that represents the asynchrony of a `trackedFunction`
  */
 export class State<Value> {
-  @tracked isResolved = false;
+  @tracked isSuccessful = false;
+  @tracked isError = false;
   @tracked resolvedValue?: Value;
   @tracked error?: unknown;
+  @tracked previousResolvedValue?: Value;
 
   #fn: ResourceFn<Value>;
   #hooks: Hooks;
   #initialValue: Value | undefined;
 
-  constructor(fn: ResourceFn<Value>, hooks: Hooks, initialValue?: Value) {
+  constructor(
+    fn: ResourceFn<Value>,
+    hooks: Hooks,
+    initialValue?: Value,
+    previousResolvedValue?: Value
+  ) {
     this.#fn = fn;
     this.#hooks = hooks;
     this.#initialValue = initialValue;
+    this.previousResolvedValue = previousResolvedValue;
+    this.retry();
   }
 
   get value() {
     return this.resolvedValue || this.#initialValue || null;
   }
 
   get isPending() {
-    return !this.isResolved;
+    return !this.isError && !this.isSuccessful;
   }
 
   get isLoading() {
     return this.isPending;
   }
 
-  get isError() {
-    return Boolean(this.error);
+  get isFinished() {
+    return !this.isPending;
   }
 
+  get previousValue() {
+    return this.previousResolvedValue || this.#initialValue || null;
+  }
+
+  protected _retry = async () => {
+    try {
+      // We need to invoke this before going async so that tracked properties are consumed (entangled with) synchronously
+      const notQuiteValue = this.#fn(this.#hooks);
+
+      // Start a new JS thread to avoid the "modified twice in a single render" error
+      await new Promise((r) => setTimeout(r, 0));
+
+      if (this.resolvedValue !== undefined) {
+        this.previousResolvedValue = this.resolvedValue;
+      }
+
+      this.isSuccessful = false;
```

Comment:
shouldn't isSuccessful be a getter based on not having an error?

Reply:
I initially made the `resolvedValue` and `error` properties the source of truth: when one of them stops being undefined, then the promise has finished.

The problem is that the promise can resolve with `undefined` which will result in an inconsistent state of `State`. Yes, the JSON spec does not have `undefined`, but network requests aren't the only use case for `trackedFunction` and `State`. The developer can choose to `resolve()` or `reject()` their promise without arguments, which resolves/rejects with `undefined`. That's legit JavaScript, so `State` should support it.

Thus, I had to make state flags simple properties. Since I went with boolean properties, and there are three unique states (pending/successful/failed), I had to employ two boolean properties.

An alternative would be to have a single simple property to store promise state, e. g.:

```ts
type PromiseState = 'pending' | 'successful' | 'error';
```

or

```ts
enum PromiseState { Pending, Successful, Error }
```

Then `isSusccessful` and `isError` can be getters.

***

Please tell me which option you prefer.

Reply:
Yeah, this makes sense -- and now I'm kinda wondering if we need to approach `State` differently, like define a stricter set of states that `State` can be in, like:

```ts
type InternalState = 
 | {
     status: 'pending';
   }
 | {
     status: 'error';
     error: Error;
   }
 | {
   status: 'successful';
   value: /* whatever this would be */
   }  
```

Then, we can be reasonably certain there is no possible overlap between the InternalStates?
thoughts?

Reply:
your idea of implementing the `State` class is 💯 I only mean that the return type be different -- rather than exposing the class as the return type, we annotate the return type to be this `InternalState`